### PR TITLE
[core] Do not overwrite session keys when pushing to the key value store during start

### PR DESCRIPTION
**Pull request by vitsai**
Right now, we overwrite various session keys during a restart. If these values are persisted, we should just directly reuse them.

## Related issue number

https://github.com/ray-project/ray/issues/38796

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/_private/node.py [hunk 1]
```diff
@@ -1178,26 +1177,26 @@ def _write_cluster_info_to_kv(self):
         self.get_gcs_client().internal_kv_put(
```
Comment (iycheng) [inline python/ray/_private/node.py:1177]: I think we should also fetch it from GCS in case it's not overwritten.

**Comment by rkooo567** [comment]
This will still have the wrong session id right? the node.py generates the session id

**Review by larrylian — APPROVED** [review]

**Comment by larrylian** [comment]
1. I think this issue need a more detailed design is required.  

a. self._session_name is generated in first.
b. Then init tmp dir by `self._session_name` in _init_temp(self).
c. Then start gcs server
d.  If not overwirte "session_name"  in `self._write_cluster_info_to_kv` ,  it may result in inconsistency between 'tmp_dir' and the 'session_name' in Redis, potentially causing certain issues.

```
None.__init__():
...
 if head:
        # date including microsecond
        date_str = datetime.datetime.today().strftime("%Y-%m-%d_%H-%M-%S_%f")
        self._session_name = f"session_{date_str}_{os.getpid()}"

....
self._init_temp()
...

self.start_head_processes()
        self.start_gcs_server()
        assert self.get_gcs_client() is not None
        self._write_cluster_info_to_kv()
```


2. We should update self._session_name var and others keys in python , if these keys are persisted.   These variables may be used in the subsequent process.

3. We should print log , if these keys are persisted and resue them.

File: python/ray/_private/node.py [hunk 2]
```diff
@@ -1175,38 +1193,18 @@ def _write_cluster_info_to_kv(self):
...
-            True,
-            ray_constants.KV_NAMESPACE_SESSION,
-        )
-        self.get_gcs_client().internal_kv_put(
-            b"temp_dir",
-            self._temp_dir.encode(),
-            True,
-            ray_constants.KV_NAMESPACE_SESSION,
-        )
+        self._get_or_set_value(b"session_name", self._session_name)
```
Comment (larrylian) [inline python/ray/_private/node.py:1196]: 1. We should update self._session_name variable and  variables of others keys in python , if these keys are persisted.   These variables may be used in the subsequent process.

```
self._session_name
... ohters
```

2. We should print log , if these keys are persisted and resue them.

**Comment by vitsai** [comment]
@larrylian Yes. init tmp dir logic needs to move to after `_write_cluster_info_to_kv`, once the proper values have been fetched from GCS.

There is also the issue of session_name being used in initialization of stats collection in `gcs_server_main` before the kv client has been initialized (so old session_name cannot be fetched). It's not clear that we can move the initialization of the stats object to a later point, though if we can that would be the easiest fix. 

Resetting session_name later (once it has been fetched from the kv store) is not super ideal since the StatsConfig object is not thread-safe and the tags in particular are quite read-heavy, so we probably wouldn't want to serialize on a mutex.

**Comment by vitsai** [comment]
Yes, I will add tests when the change is finalized. To ensure consistent state of the session keys with Redis, a few more things have to change (currently in the process of completing these):

1. temp dir, storage, and cached port initialization will move to after the initialization of GCS, but before the other head processes start
2. GCS server port will never be retrieved from cache (though it doesn't look like it ever was going through the retrieval logic anyway), and will be set before starting GCS, and cached after starting GCS but before retrieving/setting the other cached ports.
3. stats initialization in GCS server itself will probably be pushed to later; in particular, the session name and dir keys must be retrieved from Redis before they can be used for initialization. (basically need to check the passed-in value with the persisted one before using) there is also the less clean option of updating post-initialization which could work for session name but not so much for log dir.
4. gcs_table_storage to add some KV functionality (eventually the two should be merged anyway) to facilitate retrieving the keys during Init rather than during Start (so that stats initialization does not have to be pushed down very far).

Most of these have to do with session_name, session_dir, and temp_dir being used both before and after the initial point of persisting to the KV store; now that point has become a sync point with existing values in the KV store, so anything using those values has to be reordered to after

**Comment by vitsai** [comment]
Synced offline with @rkooo567 and @iycheng. This approach is not viable due to the complexity required to break the circular dependency of the following: 

1. create/fetch session name
3. create session dir tmp dir and log dirs and stderr stdout files using session name, 
3. start GCS (requires log_dir, stdout file, stderr file)

We would bring in a lot of edge cases and custom logic for it.

**Pull request closed without merging**